_posts/2021-02-26-data-validation.md

```diff
@@ -0,0 +1,195 @@
+title: Data validation
+permalink: /:title:output_ext
+layout: post
+date: 2021-02-26
+image: data-validation.jpg
+alt: Data validation
+---
+
+# Data validation
+
+![Data validation](assets/img/posts/data-validation.jpg)
+[@heapdump](https://unsplash.com/@heapdump)
+
+## Domain validation
+In my previous [blog post](http://arnolanglade.github.io/command-handler-patterns.html), I spoke about command and command handler design patterns. I got several questions about data validation and how to give feedback to users. Let’s take the example of my previous blog post: an account creation. To create an account, my business expert expects that I provide a username with at least three characters and a password with at least eight characters, an uppercase letter, a lowercase letter, and a number.
+
+How to make sure the username and password will match the rules given by our business expert? Value object will help us to achieve that. I **strongly recommend you** to wrap those properties into value objects which cannot be instantiable if data provided by users are wrong. It is a good way to introduce new types in your codebase and make it clearer and business-focused.
+
+```php
+final class Username
+{
+    private string $username;
+
+    public function __construct(string $username)
+    {
+        if (\strlen($username) < 3) {
+            throw new \InvalidArgumentException('The username is too short, it should contain at least 3 characters');
+        }
+
+        $this->username = $username;
+    }
+}
+
+final class Password
+{
+    private string $password;
+
+    public function __construct(string $password)
+    {
+        if (1 !== \Safe\preg_match('#(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{8,}#', $password)) {
+            throw new \InvalidArgumentException(
+                'The password must contain at least 8 characters, an uppercase letter, lowercase letter and a number'
+            );
+        }
+
+        $this->password = $password;
+    }
+}
+```
+
+Then we are going to modelize an account aggregate with username and password properties.
+
+```php
+final class Account
+{
+    private Username $username;
+    private Password $password;
+
+    public function __construct(
+        Username $username,
+        Password $password
+    ) {
+        $this->username = $username;
+        $this->password = $password;
+    }
+}
+```
+
+## Superficial validation
+Thanks to the domain validation we will be able to give feedback to users if they provide invalid data. As I explained in my previous [blog post](http://arnolanglade.github.io/command-handler-patterns.html) we create a command `CreateAnAccount` to represent the account creation user intent. I will show you how to use those value objects to validate command data. For that, we will use the Symfony validator component, and to be more precise we will only use the callback constraint. First, thanks to annotation will configure the validator to call a static method to validate command properties.
```

Comment:
Thanks to the "domain" validation

ton titre parle de superficial, c'est normal